compiler/src/iree/compiler/Dialect/HAL/Utils/ExternBuildingUtils.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef IREE_COMPILER_DIALECT_HAL_UTILS_EXTERN_BUILDING_UTILS_H_
+#define IREE_COMPILER_DIALECT_HAL_UTILS_EXTERN_BUILDING_UTILS_H_
+
+#include "iree/compiler/Dialect/HAL/Utils/ExternBuildingUtils.h"
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "mlir/Dialect/PDLInterp/IR/PDLInterp.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/IRMapping.h"
+#include "mlir/IR/Location.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace HAL {
+
+namespace {
+
+/// Helper to build a hal.dispatch.extern op with the given arguments. This
+/// returns the block arguments of the workgroup count region and the extern
+/// op. Note that the workgroup count region will not include the terminator
+/// and that is left up to the user to properly populate.
+static FailureOr<Operation *>
+createDispatchExtern(PatternRewriter &rewriter, ValueRange workload,
```

Comment:
this can be moved to a builder on the op (no need for the PatternRewriter here - a normal OpBuilder would do as passed to the builders)